src/amo/utils/addons.js

```diff
@@ -112,24 +112,35 @@ export const getAddonJsonLinkedData = ({
   });
 };
 
-export const getPromotedCategory = ({
+export const getPromotedCategories = ({
   addon,
   clientApp,
   forBadging = false,
 }: {|
   addon: AddonType | CollectionAddonType | SuggestionType | null | void,
   clientApp: string,
   forBadging?: boolean,
-|}): PromotedCategoryType | null => {
-  let category = null;
-  if (addon && addon.promoted && addon.promoted.apps.includes(clientApp)) {
-    category = addon.promoted.category;
+|}): Array<PromotedCategoryType> => {
+  let categories: Array<PromotedCategoryType> = [];
+  const promoted = addon?.promoted;
```

Comment:
I'm confused about the if/else statement. Why aren't we checking the prom.category truthy-ness when promoted is an array? Is that a different in behavior on the API level?

An idea of how you could prepare the array to simplify the check conditions.. assuming that diff is not expected. If we should in fact only check the category based on the value of the underlying data please add a comment explaining why that is.

```js
// return if undefined to make typechecks happy
if (!addon?.promoted) return [];


// initialize an array so the loop is simpler
const promoted = Array.isArray(addon.promoted) ? addon.promoted : [addon.promoted];
// now categories is just a filtered array of the original promoted array
// should we also check if promot.category is truthy? BTW?
let categories: Array<PromotedCategoryType> = promoted
  .filter((promo) => promo.apps.includes(clientApp));

// similar to a comment above, negated loops are confusing.
if (forBadging && categories.every((item) => !BADGE_CATEGORIES.includes(item))) {
   categories = [];
}
```

You could collapse the whole loop down to the categories variable also

```js
let categories = (
  Array.isArray(addon.promoted) ? addon.promoted : [addon.promoted]
).filter((promo) => promo.apps.includes(clientApp));
```

I think keeping the loop is good for readability later but since you now have an array of promoted categories it is very simple the rest of the logic.

Reply:
> I'm confused about the if/else statement. Why aren't we checking the prom.category truthy-ness when promoted is an array? Is that a different in behavior on the API level?

Typescript is specifically expecting addons.promoted to be an array of type Array<PromotedType>, so it being an Array -> assumes promo has the category. Which was different from the else section, since it could be of type Promoted | null. I guess we could do the check if we want to suppress that runtime error? But I don't see any reason to given that category not existing in the Array item would, well, **be** an error.

It's also why the filter doesn't work, since we're returning an array of PromotedCategoryType, but the promoted const is of type PromotedType. Everything else makes sense though.